core/chains/evm/logpoller/log_poller.go

```diff
@@ -879,9 +911,49 @@ func (lp *logPoller) PollAndSaveLogs(ctx context.Context, currentBlockNumber int
 	}
 }
 
+// Get latest and update finalityDepth
+func (lp *logPoller) latestBlockAndFinalityDepth(ctx context.Context) (*evmtypes.Head, int64, error) {
+	// If finality is not enabled, we can just get the latest block.
+	if !lp.useFinalityTag {
+		latestBlock, err := lp.ec.HeadByNumber(ctx, nil)
+		return latestBlock, lp.finalityDepth, err
+	}
+
+	// If finality is enabled, we need to get the latest and finalized blocks.
+	latest := evmtypes.Head{}
+	finalized := evmtypes.Head{}
+	batch := []rpc.BatchElem{
+		{
+			Method: "eth_getBlockByNumber",
+			Args:   []interface{}{rpc.LatestBlockNumber.String(), false},
+			Result: &latest,
+		},
+		{
+			Method: "eth_getBlockByNumber",
+			Args:   []interface{}{rpc.FinalizedBlockNumber.String(), false},
+			Result: &finalized,
+		},
+	}
+	if err := lp.ec.BatchCallContext(ctx, batch); err != nil {
+		return nil, 0, err
+	}
+
+	// All requests have to succeed to properly update finalityDepth.
+	for _, req := range batch {
+		if req.Error != nil {
+			return nil, 0, req.Error
+		}
+	}
+
+	lp.finalityDepthMu.Lock()
+	lp.finalityDepth = latest.Number - finalized.Number
+	lp.finalityDepthMu.Unlock()
+	return &latest, latest.Number - finalized.Number, nil
+}
+
 // Find the first place where our chain and their chain have the same block,
 // that block number is the LCA. Return the block after that, where we want to resume polling.
-func (lp *logPoller) findBlockAfterLCA(ctx context.Context, current *evmtypes.Head) (*evmtypes.Head, error) {
+func (lp *logPoller) findBlockAfterLCA(ctx context.Context, current *evmtypes.Head, finalityDepth int64) (*evmtypes.Head, error) {
```

Comment:
Perhaps worth noting that this makes the lp _much_ more resilient for chains with finalityTags. If a chain is producing blocks but still failing to finalize for whatever reason, we'll now automatically adjust the potential reorg depth appropriately. That will prevent the optimism test outages we've seen twice so far (reorg greater than fixed depth)